sandbox/grist/dropdown_condition.py

```diff
@@ -1,10 +1,72 @@
+import ast
+import asttokens
 import json
 import logging
+import textbuilder
 
-from predicate_formula import parse_predicate_formula_json
+from predicate_formula import NamedEntity, parse_predicate_formula_json, TreeConverter
+import predicate_formula
 
 log = logging.getLogger(__name__)
 
+class _DCEntityCollector(TreeConverter):
+  def __init__(self):
+    self.entities = []
+
+  def visit_Attribute(self, node):
+    parent = self.visit(node.value)
+
+    if parent == ["Name", "choice"]:
+      self.entities.append(NamedEntity("choiceAttr", node.last_token.startpos, node.attr, None))
+    elif parent == ["Name", "rec"]:
+      self.entities.append(NamedEntity("recCol", node.last_token.startpos, node.attr, None))
+
+    return ["Attr", parent, node.attr]
+
+
+def perform_dropdown_condition_renames(useractions, renames):
+  """
+  Given a dict of column renames of the form {(table_id, col_id): new_col_id}, applys updates
+  to the affected dropdown condition formulas.
+  """
+  updates = []
+
+  for col in useractions._engine.docmodel.columns.all:
+
+    patches = []
+
+    # Find all columns in the document that has dropdown conditions.
+    try:
+      widget_options = json.loads(col.widgetOptions)
+      dc_formula = widget_options["dropdownCondition"]["text"]
+    except (json.JSONDecodeError, KeyError):
+      continue
+
+    # Find out what table this column refers to and belongs to.
+    ref_table_id = col.type.lstrip("Ref:")
```

Comment:
Something to watch out for here: Reference List columns have a different prefix (`RefList:`).